src/nodes/chunk_insert_state.c

```diff
@@ -566,7 +565,18 @@ ts_chunk_insert_state_create(const Chunk *chunk, ChunkDispatch *dispatch)
 						 "insert into a compressed chunk that has primary or unique constraint is "
 						 "not supported")));
 		}
-		compress_rel = table_open(compress_chunk_relid, RowExclusiveLock);
+		/* fetch the chunk again after acquiring lock and verify if it is
+		 * still compressed. Then get the correct compressed chunk id
+		 * ( github issue 3400)
+		 */
+		Chunk *orig_chunk = ts_chunk_get_by_id(chunk->fd.id, true);
```

Comment:
The function `ts_chunk_insert_state_create()` uses a lot of `chunk->` pointers before, after, and outside the `if (has_compressed_chunk)` condition. Even the result of the `if (has_compressed_chunk)` condition cannot be trusted as it has been calculated before taking the lock.

In my opinion the function should be refactored so that its arguments don't have a `const Chunk *chunk` pointer at all. This should be possible to do with 2 arguments, the table id and the chunk id. The table id will allow to take the lock and the chunk id will allow to get the `Chunk` pointer after the lock has been taken. Taking the lock and the `Chunk` pointer should be the first things this function does.

Reply:
the new lines of code have been moved to a separate function.